src/qibolab/instruments/emulator/backends/generic.py

```diff
@@ -0,0 +1,315 @@
+from collections import OrderedDict
+from typing import List, Optional, Union
+
+import matplotlib.pyplot as plt
+import numpy as np
+from IPython.display import Latex, display
+from matplotlib.figure import Figure
+
+
+def dec_to_basis_string(x: int, nlevels: list = [2]) -> list:
+    """Converts an integer to a generalized bitstring in the computation basis
+    of the full Hilbert space.
+
+    Args:
+        x (int): The integer (decimal number) to be converted.
+        nlevels (list): The list of nlevels to convert the decimal number to. Defaults to [2].
+
+    Returns:
+        list: Generalized bitstring of x.
+    """
+    nqubits = len(nlevels)
+    output_list = []
+    y = x
+    subdims_ = np.multiply.accumulate(nlevels)
+    subdims = (subdims_[-1] / subdims_).astype(int)
+
+    for sub_dim in subdims:
+        coeff = np.divmod(y, sub_dim)[0]
+        output_list.append(coeff)
+        y -= coeff * sub_dim
+
+    return output_list
+
+
+def make_comp_basis(
+    qubit_list: List[Union[int, str]], qid_nlevels_map: dict[Union[int, str], int]
+) -> np.ndarray:
+    """Generates the computational basis states of the Hilbert space.
+
+    Args:
+        qubit_list (list): List of target qubit indices to generate the local Hilbert space of the qubits that respects the order given by qubit_list.
+        qid_nlevels_map (dict): Dictionary mapping the qubit IDs given in qubit_list to their respective Hilbert space dimensions.
+
+    Returns:
+        `np.ndarray`: The list of computation basis states of the local Hilbert space in a numpy array.
+    """
+    nqubits = len(qubit_list)
+
+    qid_list = [str(qubit) for qubit in qubit_list]
+    nlevels = [qid_nlevels_map[qid] for qid in qid_list]
+    comp_basis_list = []
+    comp_dim = np.prod(nlevels)
+    for ind in range(comp_dim):
+        comp_basis_list.append(dec_to_basis_string(ind, nlevels=nlevels))
+
+    return np.array(comp_basis_list)
```

Comment:
I take it back: if you can avoid generating the whole Cartesian product, it would be only better: just use integers in `[0, np.prod(nlevels))` range as the state labels, and then use `dec_to_basis_string` (possibly vectorized) only when strictly needed.

Reply:
Just to clarify, so no change is needed?

Reply:
ok in that case I'll leave this as it is now and KIV for now.

Reply:
> KIV

Interesting...
https://en.wikipedia.org/wiki/List_of_Singapore_abbreviations#K

xD

Reply:
I apologize, I assumed it was a global acronym xD

Reply:
Simplified `make_comp_basis` in d8b2ebd32e5d0bc392ee02eab54875ac842500ab.

Reply:
The function is now optimal. I knew it could have been simpler with NumPy, but I was unaware of `np.indices` function: it's exactly what you needed.

If you don't mind, I would still keep this conversation open for another while. There is only one option that could be even better: don't generate it all. But I'm not sure that's even true.

I believe you're using that more or less as a function `n: (index := [i0, i1, ...])`. If you always need to compute all the values of this function, what you're doing is already optimal, because you're essentially precomputing it in the optimal way, and then looking up (you should avoid only if you're exhausting memory, it's a space-time tradeoff).
However, if you can avoid using all the values, because you can evaluate on-demand (following end-user requests), then we could save some evaluations.

However, it's already more than fine as it is, I just want to give it a second thought later on.